comet/broker.py

```diff
@@ -409,8 +411,19 @@ async def register_dataset(request):
 
         # Lack datasets and check if dataset already known.
         async with lock_datasets:
-            found = await get_dataset(hash, wait=False)
-            if found is not None:
+            try:
+                found = await get_dataset(hash, wait=False)
+            except DatasetNotFoundError:
+                if dataset_valid and root is not None:
+                    # save the dataset
+                    await redis.execute("hset", "datasets", hash, json.dumps(ds))
+
+                    reply["result"] = "success"
+                    archive_ds = True
+
+                    # Notify anything waiting for this dataset to arrive
+                    signal_created(hash, "dataset", lock_datasets, waiting_datasets)
```

Comment:
What should happen if you get `DatasetNotFoundError` but you don't satisfy the `if` clause above? I appreciate this isn't covered in the original code either.

Reply:
result gets set to "dataset invalid", this branch gets skipped and the function returns. I agree it's not too readable.